compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -726,10 +726,9 @@ class TypeComparer(initctx: Context) extends DotClass with ConstraintHandling {
 
     tycon2 match {
       case param2: TypeParamRef =>
-        isMatchingApply(tp1) || {
-          if (canConstrain(param2)) canInstantiate(param2)
-          else compareLower(bounds(param2), tyconIsTypeRef = false)
-        }
+        isMatchingApply(tp1) ||
+        canConstrain(param2) && canInstantiate(param2) ||
```

Comment:
Just merged #2204. I don't have a separate test case for this change, but would argue that the fact that we missed the case before means we cannot exclude that we will miss it in the future. So better play it safe.